packages/react-components/src/components/TextFieldWithMention/TextFieldWithMention.tsx

```diff
@@ -0,0 +1,629 @@
+            );
+          }
+          setSelectionStartValue(mentionTag.plainTextBeginIndex);
+          setSelectionEndValue(mentionTag.plainTextEndIndex ?? mentionTag.plainTextBeginIndex);
+        } else {
+          setSelectionStartValue(currentSelectionStart);
+          setSelectionEndValue(currentSelectionEnd);
+        }
+      } else {
+        // selection was changed by keyboard
+        updateSelectionIndexesWithMentionIfNeeded(event, inputTextValue, selectionStartValue, selectionEndValue, tags);
+      }
+      // don't set setShouldHandleOnMouseDownDuringSelect(false) here as setSelectionRange could trigger additional calls of onSelect event
+      // and they may not be handled correctly (because of setSelectionRange calls or rerender)
+    },
+    [updateSelectionIndexesWithMentionIfNeeded]
+  );
+
+  const updateCurrentTriggerStartIndexAndQuery = useCallback(
+    async (
+      newValue: string,
+      triggerText: string,
+      currentSelectionEndValue: number,
+      event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>,
+      currentTriggerStartIndex: number
+    ) => {
+      // If we are enabled for lookups,
+      if (mentionLookupOptions !== undefined) {
+        // Look at the range of the change for a trigger character
+        const triggerPriorIndex = newValue.lastIndexOf(triggerText, currentSelectionEndValue - 1);
+        // Update the caret position, if not doing a lookup
+        setCaretPosition(Caret.getRelativePosition(event.currentTarget));
+        if (triggerPriorIndex !== undefined) {
+          // trigger is found
+          const isSpaceBeforeTrigger = newValue.substring(triggerPriorIndex - 1, triggerPriorIndex) === ' ';
+          const wordAtSelection = newValue.substring(triggerPriorIndex, currentSelectionEndValue);
+          let tagIndex = currentTriggerStartIndex;
+          if (!isSpaceBeforeTrigger && triggerPriorIndex !== 0) {
+            //no space before the trigger <- continuation of the previous word
+            tagIndex = -1;
+            setCurrentTriggerStartIndex(tagIndex);
+          } else if (wordAtSelection === triggerText) {
+            // start of the mention
+            tagIndex = currentSelectionEndValue - triggerText.length;
+            if (tagIndex < 0) {
+              tagIndex = 0;
+            }
+            setCurrentTriggerStartIndex(tagIndex);
+          }
+          if (tagIndex === -1) {
+            await debouncedQueryUpdate(undefined);
+          } else {
+            // In the middle of a @mention lookup
+            if (tagIndex > -1) {
+              const query = wordAtSelection.substring(triggerText.length, wordAtSelection.length);
+              if (query !== undefined) {
+                await debouncedQueryUpdate(query);
+              }
+            }
+          }
+        }
+      }
+    },
+    [debouncedQueryUpdate, mentionLookupOptions]
+  );
+
+  const getUpdatedNewValueForOnChange = useCallback(
```

Comment:
Is all the value consumed by this function all coming from parameter? if yes, would like to see a separated functions for this

Reply:
I mean a const function not living in the component function body

Reply:
Yep, no dependencies array for useCallback would be a good indicator that this function might be able to be moved outside of component function body

Reply:
> function

Moved it out to the mentionTagHelpers.ts